controllers/machine_controller_noderef.go

```diff
@@ -170,29 +169,41 @@ func summarizeNodeConditions(node *corev1.Node) (corev1.ConditionStatus, string)
 
 func (r *MachineReconciler) getNode(ctx context.Context, c client.Reader, providerID *noderefutil.ProviderID) (*corev1.Node, error) {
 	log := ctrl.LoggerFrom(ctx, "providerID", providerID)
-
 	nodeList := corev1.NodeList{}
-	for {
-		if err := c.List(ctx, &nodeList, client.Continue(nodeList.Continue)); err != nil {
-			return nil, err
-		}
+	if err := c.List(ctx, &nodeList, client.MatchingFields{clusterv1.NodeProviderIDIndex: providerID.IndexKey()}); err != nil {
+		return nil, err
+	}
+	if len(nodeList.Items) == 0 {
+		// If for whatever reason the index isn't registered or available, we fallback to loop over the whole list.
+		nl := corev1.NodeList{}
+		for {
+			if err := c.List(ctx, &nl, client.Continue(nl.Continue)); err != nil {
+				return nil, err
+			}
+
+			for key, node := range nl.Items {
+				nodeProviderID, err := noderefutil.NewProviderID(node.Spec.ProviderID)
+				if err != nil {
+					log.Error(err, "Failed to parse ProviderID", "node", client.ObjectKeyFromObject(&nl.Items[key]).String())
+					continue
+				}
 
-		for _, node := range nodeList.Items {
-			nodeProviderID, err := noderefutil.NewProviderID(node.Spec.ProviderID)
-			if err != nil {
-				log.Error(err, "Failed to parse ProviderID", "node", node.Name)
-				continue
+				if providerID.Equals(nodeProviderID) {
+					return &node, nil
+				}
 			}
 
-			if providerID.Equals(nodeProviderID) {
-				return &node, nil
+			if nl.Continue == "" {
+				break
 			}
 		}
 
-		if nodeList.Continue == "" {
-			break
-		}
+		return nil, ErrNodeNotFound
+	}
+
+	if len(nodeList.Items) > 1 {
```

Comment:
Can't `nodeList` still be zero at this point?

Reply:
it shouldn't be, if it was 0 before and we entered the for loop we should have returned either the node or `ErrNodeNotFound` if it wasn't found

Reply:
for safety this can be 

```suggestion
	if len(nodeList.Items) != 1 {
```

though

Reply:
Added a test.